futures/src/lib.rs

```diff
@@ -23,126 +25,18 @@
 
 #![no_std]
 
-#![deny(missing_docs, missing_debug_implementations, warnings)]
+#![deny(missing_docs, missing_debug_implementations)]
 #![deny(bare_trait_objects)]
 
-#![doc(html_root_url = "https://docs.rs/futures/0.2.0")]
+#![doc(html_root_url = "https://docs.rs/futures-preview/0.3.0-alpha")]
 
 #![cfg_attr(feature = "nightly", feature(cfg_target_has_atomic))]
 #![cfg_attr(feature = "nightly", feature(use_extern_macros))]
 
-// extern crate futures_async_runtime;
-extern crate futures_core;
-extern crate futures_channel;
-extern crate futures_executor;
-extern crate futures_io;
-extern crate futures_sink;
-extern crate futures_util;
-
-#[doc(hidden)]
-pub use futures_core::core_reexport;
-pub use futures_core::future::Future;
-pub use futures_util::future::FutureExt;
-pub use futures_core::stream::Stream;
-pub use futures_util::stream::StreamExt;
-pub use futures_sink::Sink;
-pub use futures_util::sink::SinkExt;
```

Comment:
I removed these top level exports because using them is an anti-pattern. It's much easier to either import from the right location like `futures::task::Poll` or to import the prelude `futures::prelude::*`. Anything in-between doesn't make sense.

Reply:
I disagree. I don't see any anti-pattern. I frequently have `use futures::{Future, Stream}` in modules. Making them harder to import doesn't improve anything.

Glob imports and preludes, however, definitely *are* an anti-pattern.

Reply:
I agree with @seanmonstar, having to begin every file with

```rust
use futures::{core::{future::Future, stream::Stream}, utils::{future::FutureExt, stream::StreamExt}};
```

Instead of

```rust
use futures::{Future, Stream, FutureExt, StreamExt};
```

Would be very annoying.

Reply:
@Nemo157 @seanmonstar Thanks for your comments!

My reaction is that all this top-level stuff accumulates. There is also:
```
Context, ContextExt, CoreFutureExt, FutureObj, LocalFutureObj
```
These are just the most common imports, but there are obviously more. (`FutureObj` exists because `dyn Trait` is currently impossible for traits that use arbitrary self types)  My intention behind removing these top-level imports is because it makes the docs better. Top level exports are a bad thing because they destroy the structure. If everything is exported at the top level anyway, there is no real reason to have submodules at all.  How can we decide which things are considered important enough to be exported at the top level? How will it grow in the future? How would it look like if the standard library did it? I think that it doesn't scale and we shouldn't do it.
The hierarchy I'm proposing here is the same as in libcore: libcore also has differentiates between the submodules `future` and `task`. And a `stream` submodule will probably exist in the future.

@Nemo157 `FutureExt` is defined in the utils crate, but there's there's no such module

This is how I imagine it:
```rust
use futures::future::{Future, FutureExt};
use futures::stream::{Stream, StreamExt};
use futures::task::{Task, Context};
```
I think that actually looks pretty nice. That said, the approach I'm proposing here is just a suggestion. This PR has become larger large, so I explictly highlighted this change to make it absolutely obvious. I certainly don't want to decide that for everyone. If you all really dislike it, we can revert to the old approach. I'm confident, though, that the approach of clear separation is better.

Reply:
I think idea of top-level re-exports serves both convenience for common case (easy to reach things you need without clutter) and discoverability. It's hard to argue that Future, Stream and Sink (and their extension traits since 0.2) are core concepts to futures crates. In my view, public submodules still make sense for things that are more specialized, e.g. deeper dive into streaming functionality for writing your own combinators. Being able to import things through `futures::{Future, FutureExt, Stream, StreamExt}` and `futures::stream::{Stream, StreamExt, FilterMap, Empty}` is both useful in my opinion if only in different contexts.

Reply:
Clean:
<img width="635" alt="screen shot 2018-07-09 at 17 32 27" src="https://user-images.githubusercontent.com/340142/42460432-2190b47e-839e-11e8-8ac2-b5ccb583e8a3.png">

Cluttered:
<img width="609" alt="screen shot 2018-07-09 at 17 38 22" src="https://user-images.githubusercontent.com/340142/42460732-ed4e74e8-839e-11e8-897a-2bdfdecd2e60.png">

Reply:
I don't see any problem with the traits list there at the bottom. *If* it were a big deal, they can be `doc(hidden)` from the main page, but I don't think it's an issue.

I definitely wouldn't make actually *using* these traits harder just for the docs change.

Reply:
One thing to note is that `futures` 0.2 didn't export `Context` to the top-level. As far as I am aware the standard was to use `task::Context` everywhere. That doesn't reduce it much though 😦

I wonder if trait aliases could help with all those extension traits. In `futures` have something like

```rust
trait Future = ::core::future::Future + CoreFutureExt + FutureExt;
```

(although currently the RFC doesn't specify that trait aliases bring methods into scope, [I left a comment on the tracking issue to find out](https://github.com/rust-lang/rust/issues/41517#issuecomment-403550419)).

Reply:
@seanmonstar My main concern is the impression a new user gets when looking at the docs for the first time. I've added a commit that adds the reexports as `doc(hidden)`. As long as there is no big pile of unsorted stuff visible to users, it's fine I guess.

@Nemo157 I think it would be confusing if more than one thing was called a `Future`. Would `impl Future for MyType` even work for a trait alias?